compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUOps.cpp

```diff
@@ -203,8 +203,26 @@ LogicalResult MultiMmaOp::verify() {
   int64_t accInnerElementCount = multiplyAcc(getAccInnerShape());
 
   auto [m, n, k] = getKind().getMNKShape();
-  if (m * k != lhsInnerElementCount || n * k != rhsInnerElementCount ||
-      m * n != accInnerElementCount) {
+  int64_t expectedNumLhsElem = m * k;
+  int64_t expectedNumRhsElem = n * k;
+  int64_t expectedNumAccElem = m * n;
+
+  // The below variables have zero values when the op is at thread level.
+  int64_t M0K0 = lhsInnerElementCount / m / k; // (M0M1K0K1) / M1 / K1 = M0K0
+  int64_t N0K0 = rhsInnerElementCount / n / k; // (N0N1K0K1) / N1 / K1 = N0K0
+  int64_t M0N0 = accInnerElementCount / m / n; // (M0M1N0N1) / M1 / N1 = M0N0
+  if (hasPureTensorSemantics() && M0K0 && N0K0 && M0N0) {
+    int mUnrollFactor = std::sqrt(M0K0 * M0N0 / N0K0);
+    int nUnrollFactor = std::sqrt(M0N0 * N0K0 / M0K0);
+    int kUnrollFactor = std::sqrt(M0K0 * N0K0 / M0N0);
+    expectedNumLhsElem *= mUnrollFactor * kUnrollFactor;
+    expectedNumRhsElem *= nUnrollFactor * kUnrollFactor;
+    expectedNumAccElem *= mUnrollFactor * nUnrollFactor;
+  }
```

Comment:
@qedawkins I end up with making getMNKShape return intrinsic size. Because the attribute itself does not have shapes, so it really can't infer MNK shape. Instead, here is where the unrolling factor inference happens. I figured the formula to calculate the unrolling factors and documented it in the comments.

The missing implementation of interface method is `populateOperandOffsetsSizesStrides`. For those methods that we don't care in data-tiling, I just put an assertion to signal the failure.

Please take a look. We can discuss some details tomorrow. If you want to look at GPU ops/attributes changes, you can take a look at this commit: https://github.com/iree-org/iree/pull/18196/commits/72b75eb2a23b18bd252f9a01f51f1ef1fca895a6#diff-334a4fcfe53c3ad70f495b1a0b7e8d5f5f4e4181e04fd359193cbc3097cea68fR206-R221

(cc @bjacob @pashu123 @qedawkins @lialan )

Reply:
I forgot one aspect when we chatted yesterday and that is that `getMNKShape` is also used by `hasThreadSemantics`, which is required by the thread distribution pattern. Rather than allowing these unrolling factors on the inner dims (which has no meaning for any non-data tiled mma attr), I would just add a `FailureOr<bool> innerShapeHasThreadSemantics` interface method to `MmaInterfaceAttr` with a default implementation that checks the MNK shape like the current verifier does (and fails if the shape is invalid). We could use this to get stronger verification of the non-data tiled path too.